### ISSUE TITLE
Cannot initialize gocryptfs due to "WriteDirIV: open .../cipher/gocryptfs.diriv: permission denied"

### ISSUE THREAD

**Issue by Natanji** [rfjakob/gocryptfs#105]
I'm trying to set up gocryptfs on Ubuntu 14.04.1 LTS.
I downloaded the binary release 1.2.1 from here (Debian version) and tried initilaizing a folder. But I get the mentioned error:

$ cd ~
$ mkdir plain cipher
$ ./gocryptfs -init cipher
Choose a password for protecting your files.
Password: 
Repeat: 
WriteDirIV: open /home/\<username>/cipher/gocryptfs.diriv: permission denied
open /home/\<username>/cipher/gocryptfs.diriv: permission denied

In case this matters: my home folder is an NFS4 mount.

**Comment by rfjakob** [comment]
gocryptfs creates the diriv files with permissions 400 (read-only) but it
obtains a writeable file handle on creation. This should work, and it
usually does, but maybe your nfs mount does not like it? Have you tried on
/tmp?

**Comment by rfjakob** [comment]
Oh, and can you post the permissions of "cipher" directory?

**Comment by Natanji** [comment]
The permissions are drwx-----x - it's right in my home folder.
On /tmp, it does work.

I don't think NFS supports opening files that are read-only with a writeable file handle, and in fact I am trying to use gocryptfs as an alternative to ecryptfs because ecryptfs seems to be doing the same thing, resulting in the inability to access read-only files on NFS file systems - and other network shares, like sshfs, samba etc.: 
https://bugs.launchpad.net/ecryptfs/+bug/1544626
https://github.com/libfuse/sshfs/pull/17

May I ask why you create the file as read-only, but then write to it? That doesn't make any sense to me and sounds like a hack - which does fail on some instances, like this one. An easy fix should be to properly create the file with write permissions, then write to it, then change it to read-only afterwards. Can you do that for the next release?

**Comment by rfjakob** [comment]
Hmm, this usually works fine, and it is the only race-free way to create a read-only file.

Are you saying this does not work in your NFS mount?

```
umask 777
echo xxx > test.txt
```

It does work on SSHFS in my testing here.

```
ls -l
----------. 1 jakob jakob    2 25. Apr 19:14 test.txt
```

**Comment by Natanji** [comment]
Indeed. It will create the file (same permissions as you have), but doesn't put xxx into it. The file is empty. As far as I was able to find out (but your search might turn up a different result), this is expected behaviour in NFS because while local users can override permissions (when a user can create files in a directory), NFS does not support this.

**Comment by lechner** [comment]
Are you sure you are in your home directory?

WriteDirIV: open /home//cipher/gocryptfs.diriv: permission denied
open /home//cipher/gocryptfs.diriv: permission denied



On Apr 25, 2017 10:35 AM, "Natanji" <notifications@github.com> wrote:

Indeed. It will create the file (same permissions as you have), but doesn't
put xxx into it. The file is empty. As far as I was able to find out (but
your search might turn up a different result), this is expected behaviour
in NFS because while local users can override permissions (when a user can
create files in a directory), NFS does not support this.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/rfjakob/gocryptfs/issues/105#issuecomment-297107290>,
or mute the thread

**Comment by Natanji** [comment]
Yes, I merely redacted my username for privacy reasons. I replaced it with \<username> above but forgot the escape the \< so it wasn't showing due to markdown syntax seeing it as a tag... I fixed it above now.

Also, I edited my previous comment but you didn't see that in the eMail, I apologize. I wrote above: "I was wrong, I forgot to chmod 777 before looking at the file. If I do that, cat'ing test.txt will print xxx. So THAT works. Huh. Now what exactly is different about the way gocryptfs creates the file, and how the above echo and then \> commands do so?"

**Comment by rfjakob** [comment]
Ok, thanks for the update! I'll check the code what else could cause this.

**Comment by lechner** [comment]
I had a similar problem when mounting gocrypts with 'allow_other' as root.
Is the diriv owned by the same user as everything else?

**Comment by lechner** [comment]
Also, are which nfs4 security mode are you using, please?

**Comment by rfjakob** [comment]
The error message is from `ioutil.WriteFile(file, iv, 0400)` ([here](https://github.com/rfjakob/gocryptfs/blob/master/internal/nametransform/diriv.go#L74)). Code looks good.

**Comment by Natanji** [comment]
How can I check the security mode?
The files in the directory are all owned by the same user as everything
else, yes. I'm not running the commands as root, but with the same user
that they belong to.


Am 25.04.2017 um 21:24 schrieb Felix Lechner:
> Also, are which nfs4 security mode are you using, please?
>

**Comment by lechner** [comment]
Do you use kerberos security (krb5i or krb5p) in /etc/exports?

**Comment by Natanji** [comment]
Yes, I do, more precisely:
gss/krb5i(rw,async,no_subtree_check,no_all_squash)

Why, are there some specifics of kerberos that I'm not aware of?

**Comment by rfjakob** [comment]
I wonder if this is actually caused by selinux or apparmor or something.
When the "echo" works gocryptfs should work as well

**Comment by Natanji** [comment]
SElinux is not installed on the local machine, and while apparmor is installed and enabled I have confirmed with aa-notify that running gocryptfs doesn't lead to any denials.

Neither SElinux nor AppArmor are installed on the fileserver from which the nfs4 is exported.

So that's probably not it.

**Comment by lechner** [comment]
I have a setup similar to yours. Here, gocryptfs runs great on top of
kerberized nfs4. It is stable and fast.

EncFS works also (but I rarely use it anymore). I do not have any of the
problems reported here <https://bugs.launchpad.net/ecryptfs/+bug/1544626>.

The differences are:

1. I track Debian testing on client and server, which may offer newer nfs
kernel daemon and other tools
2. I use gocryptfs binaries from the Debian archive, which were compiled
for my kernel

Initially, I had problems with permissions also. The nfs4 security model,
which requires a separate active kerberos ticket for each user, and its
periodic renewal, even when the machine key is configured to give the local
root user root privileges on the nfs volume, is different from a standard
unix file system. The access control in fuse adds another level of
complexity.

Will gocryptfs work for you as root after setting something like this in
/etc/idmapd.conf?

[Translation]
Method = nsswitch
GSS-Methods = static,nsswitch

[Static]
host/{FQDN}@{REALM} = root

Also, do you have the option to upgrade your nfs setup?

**Comment by Natanji** [comment]
Since this is a corporate environment, I can't change the nfs setup/risk breaking anything. I also can't update the fileserver from Debian stable.

**Comment by rfjakob** [comment]
Does it work with the WriteDirIV permission changed to 0600? (static binary attached) [gocryptfs0600.gz](https://github.com/rfjakob/gocryptfs/files/959583/gocryptfs0600.gz)

**Comment by rfjakob** [comment]
Should give you this:
```
$ ./gocryptfs0600 -init cipher
Choose a password for protecting your files.
Password: 
Repeat: 
The gocryptfs filesystem has been created successfully.
You can now mount it using: gocryptfs cipher MOUNTPOINT
0 jakob@brikett:~/go/src/github.com/rfjakob/gocryptfs$ ls -l cipher/
total 8
-r--------. 1 jakob jakob 408 26. Apr 21:49 gocryptfs.conf
-rw-------. 1 jakob jakob  16 26. Apr 21:49 gocryptfs.diriv
```

**Comment by rfjakob** [comment]
So I `strace`'d both bash and gocryptfs, they are doing things differently, which may explain why one works and the other does not:

bash:
```
umask(0777)                             = 0777
open("test.txt", O_WRONLY|O_CREAT|O_TRUNC, 0666) = 3
```
gocryptfs:
```
openat(AT_FDCWD, "/tmp/tmp.qpLwtKEF92/cipher/gocryptfs.diriv", 
    O_WRONLY|O_CREAT|O_TRUNC|O_CLOEXEC, 0400) = 3
```

**Comment by Natanji** [comment]
``` 
$ ./gocryptfs0600 -init cipher
Choose a password for protecting your files.
Password:
Repeat:
The gocryptfs filesystem has been created successfully.
You can now mount it using: gocryptfs cipher MOUNTPOINT
```
Yes, that absolutely works. I do get the same permissions (u+rw for .diriv, u+r for .conf) as you do then, so it really *is* a problem with those permissions somehow. I also tested mounting the cipher directory to a plain directory with the 0600 version, and creating a git directory within since that creates read-only files. Absolutely no problem with the 0600 version there.

What I couldn't test is whether this is just about the .diriv file or other things as well, since the latest binary release is 1.2.1 while 0600 is beta 1.3 and with the 1.2.1 release I apparently cannot mount beta 1.3 created gocryptfs directories (I'm getting the error *Unsupported feature flag "HKDF"*).

**Comment by rfjakob** [comment]
Now the plot twist is that gocryptfs.conf is also created as a read-only
file, just without using the ioutil wrapper. I'll check tonight what the
two are doing differently.

I think the possibilities of what we are looking at are:
1) bug in the nfs client. This is assuming that permissions are still
checked client-side for nfsv4. What kernel version are you running on the
client?
2) ioutil is doing something stupid or strange. I'll check tonight.

Do you want to stay on gocryptfs v1.2? v1.3 adds the HKDF step in the kry
derivation as suggested by the security audit.

On Apr 27, 2017 02:39, "Natanji" <notifications@github.com> wrote:

$ ./gocryptfs0600 -init cipher
Choose a password for protecting your files.
Password:
Repeat:
The gocryptfs filesystem has been created successfully.
You can now mount it using: gocryptfs cipher MOUNTPOINT

Yes, that absolutely works. I do get the same permissions (u+rw for .diriv,
u+r for .conf) as you do then, so it really *is* a problem with those
permissions somehow. I also tested mounting the cipher directory to a plain
directory with the 0600 version, and creating a git directory within since
that creates read-only files. Absolutely no problem with the 0600 version
there.

What I couldn't test is whether this is just about the .diriv file or other
things as well, since the latest binary release is 1.2.1 while 0600 is beta
1.3 and with the 1.2.1 release I apparently cannot mount beta 1.3 created
gocryptfs directories (I'm getting the error *Unsupported feature flag
"HKDF"*).

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/rfjakob/gocryptfs/issues/105#issuecomment-297579414>,
or mute the thread

**Comment by rfjakob** [comment]
With change 7d38f80a78644c8ec4900cc990bfb894387112ed , the `.diriv` file is created the same way the `.conf` file is. This should fix the issue without relaxing the permissions. Will be released as gocryptfs v1.3, probably this weekend.

**Comment by rfjakob** [comment]
I have released v1.3, including Debian binaries: https://github.com/rfjakob/gocryptfs/releases/tag/v1.3
